### PR TITLE
Tls add registry 01

### PR DESCRIPTION
**Pull request by chifflier**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

**Request For Comments**

This is the revival (rebase + update) of #1099 / #1136, and discuss potential common points with  #2768 
Copyright has been assigned to OISF as per usual rules.

Extracts the selected ciphersuite from the ServerHello and adds rules.
I think matching the selected ciphersuite is great, but since the ciphersuite includes lots of information, it is better to extract all of them to be able to create certain rules.
For ex, extracting the key exchange method allows to detect RSA key exchanges (See https://robotattack.org/), or we can also blacklist all weak encryption ciphers using one rule, etc.
The drawback is that the ciphersuites registry is quite big, and would be painful to maintain manually. I added python scripts to fetch the IANA XML file and transform it to the right format automatically.

My PR was quite old, and it seems there is some redundancy with #2768. It would be interesting to see what is in common and how to merge it ? cc @fooinha @thus
For ex, this PR only checks the selected (server) cipher, but the other PR also allows to extract the client proposals.

Describe changes:
- Add new TLS ciphersuites registry, containing all suites definitions and parameters, as a YAML file
- Add the scripts to create the registry from the IANA XML file, so updating the registry remains easy
- Decode the TLS ServerHello message to extract the selected ciphersuite
- Add new detection keywords. All keywords are defined as sub-keywords from tls.handshake:
    - `tls.handshake:enc=RC4` will match if encryption algorithm is RC4
    - `tls.handshake:enc-size>64` will match if encryption key size is greater
      than 64
`

`

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by victorjulien** [comment]
Other than the code implementation, I think a question is how to deal with the ciphersuites.yaml file. This is something that can be updated with new info without the need for a Suricata update AFAICS. That means it might be closer to rules than to code. While I don't mind bundling it, I think it could be an idea to start using suricata-update to keep this current as well. Then OISF would host and maintain this file and users can (through suricata-update or by other means) keep it up to date. Does that make sense?

**Comment by chifflier** [comment]
Agreed, the ciphersuites.yaml file is more data than configuration, and its format is independent of suricata (except if the format is changed). It could completely be handled by suricata-update.

**Comment by jasonish** [comment]
> Other than the code implementation, I think a question is how to deal with the ciphersuites.yaml file. This is something that can be updated with new info without the need for a Suricata update AFAICS. That means it might be closer to rules than to code. While I don't mind bundling it, I think it could be an idea to start using suricata-update to keep this current as well. Then OISF would host and maintain this file and users can (through suricata-update or by other means) keep it up to date. Does that make sense?

I propose a source that is something like "oisf/core-data". This is a source that contains some core data that may be bundled with a release, but also updated between releases. The ciphersuites file could be part of this. I'm not sure what else. But I don't want to introduce a new "source" for every bit of data we might want to update between releases, so am suggesting some "core-data" pack.

Some thought with respect to dealing with this from the engine and suricata-update side will be needed, but those are just implementation details.

File: scripts/tls/README_TLS_ciphersuites [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+:Date: 2014-03-26
+:Version: 1
+:Author: Pierre Chifflier <chifflier@wzdftpd.net>
+
+
+These scripts are used to auto-generate the YAML configuration file for
+suricata, containing the list of known TLS ciphersuites and parameters.
+
+The first script (extract-iana-ciphers.py) fetch the list from the IANA website
```
Comment (norg) [inline scripts/tls/README_TLS_ciphersuites]: *fetches*

**Comment by victorjulien** [comment]
I'm a bit confused about how to use this. I appended this line to my suricata.yaml:

include: scripts/tls/tls-ciphersuites.yaml

But then I get:

<pre>
./src/suricata  -l tmp/ -c suricata.yaml -v -T --disable-detection
[27081] 9/2/2018 -- 17:04:30 - (suricata.c:1839) <Info> (ParseCommandLine) -- Running suricata under test mode
[27081] 9/2/2018 -- 17:04:30 - (suricata.c:1714) <Info> (ParseCommandLine) -- detection engine disabled
[27081] 9/2/2018 -- 17:04:30 - (conf-yaml-loader.c:241) <Info> (ConfYamlParse) -- Including configuration file scripts/tls/tls-ciphersuites.yaml.
Error opening file /var/log/suricata/suricata.log
[27081] 9/2/2018 -- 17:04:30 - (suricata.c:1073) <Notice> (LogVersion) -- This is Suricata version 4.1.0-dev (rev 8bd763f)
[27081] 9/2/2018 -- 17:04:30 - (util-cpu.c:171) <Info> (UtilCpuPrintSummary) -- CPUs/cores online: 12
[27081] 9/2/2018 -- 17:04:30 - (conf.c:481) <Error> (ConfGetChildValueInt) -- [ERRCODE: SC_ERR_INVALID_YAML_CONF_ENTRY(139)] - malformed integer value for rfc with base 0: 'rfc5246'
[27081] 9/2/2018 -- 17:04:30 - (util-tls.c:332) <Warning> (TlsCiphersInit) -- [ERRCODE: SC_ERR_INVALID_VALUE(130)] - Could not parse cipher 0

[27081] 9/2/2018 -- 17:04:30 - (util-logopenfile.c:501) <Info> (SCConfLogOpenGeneric) -- fast output device (regular) initialized: fast.log
[27081] 9/2/2018 -- 17:04:30 - (util-logopenfile.c:501) <Info> (SCConfLogOpenGeneric) -- eve-log output device (regular) initialized: eve.json
[27081] 9/2/2018 -- 17:04:30 - (util-logopenfile.c:501) <Info> (SCConfLogOpenGeneric) -- stats output device (regular) initialized: stats.log
[27081] 9/2/2018 -- 17:04:30 - (suricata.c:2862) <Notice> (main) -- Configuration provided was successfully loaded. Exiting.
</pre>

It does appear the include works:
<pre>
./src/suricata  -l tmp/ -c suricata.yaml -v --dump-config|grep cipher|more
[30389] 9/2/2018 -- 17:04:58 - (conf-yaml-loader.c:241) <Info> (ConfYamlParse) -- Including configuration file scripts/tls/tls-ciphersuites.yaml.
tls-ciphersuites = (null)
tls-ciphersuites.0 = cipher
tls-ciphersuites.0.cipher = TLS_NULL_WITH_NULL_NULL
tls-ciphersuites.0.cs = 0x0000
tls-ciphersuites.0.name = TLS_NULL_WITH_NULL_NULL
tls-ciphersuites.0.openssl-name = NULL
tls-ciphersuites.0.kx = NULL
tls-ciphersuites.0.au = NULL
tls-ciphersuites.0.enc = NULL
tls-ciphersuites.0.enc-mode = NULL
tls-ciphersuites.0.enc-size = 0
tls-ciphersuites.0.mac = NULL
tls-ciphersuites.0.mac-size = 0
tls-ciphersuites.0.prf = DEFAULT
tls-ciphersuites.0.prf-size = 0
tls-ciphersuites.0.rfc = rfc5246
tls-ciphersuites.0.export = 0
tls-ciphersuites.0.minversion = 0x0300
tls-ciphersuites.0.maxversion = 0xffff
</pre>

**Comment by victorjulien** [comment]
@jasonish I like your suggestion. We'd ship a copy off this file in /usr/share/suricata/, then unless there is an explicit configuration first check if it exists in /var/lib/suricata/core-data/ and otherwise load it from /usr/share/suricata/
Make sense?

**Comment by victorjulien** [comment]
@chifflier are you interested in picking up #2768, remove the config stuff from it and integrating it with this PR? I think @fooinha lost interest.

**Comment by chifflier** [comment]
@victorjulien Yes, I will look at merging both PRs. It may take some time though.

About the error in the 'rfc' field, sorry for that - I updated the file, and this field changed from an integer to a string.

**Comment by regit** [comment]
@chifflier @fooinha is there a chance to revive this code ?

**Comment by regit** [comment]
> > @chifflier @fooinha is there a chance to revive this code ?
> 
> @regit The cipher suites are already decoded and added to JA3, so it would probably be easier to implement detection keywords based on that (if that's what you want/need) :)

Oh thanks, I'm gonna check that!

**Pull request closed without merging**